backends/p4tools/testgen/targets/bmv2/expr_stepper.cpp

```diff
@@ -0,0 +1,1742 @@
+             // infinite recirculation loops.
+             if (state.hasProperty("recirculate_count")) {
+                 recirculateCount = state.getProperty<uint64_t>("recirculate_count");
+                 if (recirculateCount > 1) {
+                     auto* nextState = new ExecutionState(state);
+                     ::warning("Only single recirculation supported for now. Dropping packet.");
+                     auto* dropStmt = new IR::MethodCallStatement(
+                         IRUtils::generateInternalMethodCall("drop_and_exit", {}));
+                     nextState->replaceTopBody(dropStmt);
+                     result->emplace_back(nextState);
+                     return;
+                 }
+             }
+             bool argsAreTainted = false;
+             for (size_t idx = 0; idx < args->size(); ++idx) {
+                 const auto* arg = args->at(idx);
+                 const auto* argExpr = arg->expression;
+
+                 // TODO: Frontload this in the expression stepper for method call expressions.
+                 if (!SymbolicEnv::isSymbolicValue(argExpr)) {
+                     // Evaluate the condition.
+                     stepToSubexpr(argExpr, result, state,
+                                   [call, idx](const Continuation::Parameter* v) {
+                                       auto* clonedCall = call->clone();
+                                       auto* arguments = clonedCall->arguments->clone();
+                                       auto* arg = arguments->at(idx)->clone();
+                                       arg->expression = v->param;
+                                       (*arguments)[idx] = arg;
+                                       clonedCall->arguments = arguments;
+                                       return Continuation::Return(clonedCall);
+                                   });
+                     return;
+                 }
+                 argsAreTainted = argsAreTainted || state.hasTaint(arg->expression);
+             }
+             // If any of the input arguments is tainted, the entire extern is unreliable.
+             if (argsAreTainted) {
+                 ::warning("clone args are tainted and not predictable. Skipping clone execution.");
+                 auto* nextState = new ExecutionState(state);
+                 nextState->popBody();
+                 result->emplace_back(nextState);
+                 return;
+             }
+
+             auto cloneType = args->at(0)->expression->checkedTo<IR::Constant>()->asUint64();
+             const auto* sessionIdExpr = args->at(1)->expression;
+             auto recirculateIndex = args->at(2)->expression->checkedTo<IR::Constant>()->asUint64();
+             boost::optional<const Constraint*> cond = boost::none;
+
+             if (cloneType == BMv2Constants::CLONE_TYPE_I2E) {
+                 // Pick a clone port var. For now, pick a random value from 0-511.
+                 const auto* rndConst =
```

Comment:
In some sense this is an executable specification of BMv2.
I wonder whether one could take advantage of that in some way besides test generation. For example, why do you need bmv2 at all, can you replace it with this implementation?

Reply:
Well, we can only give you the expected output for a given input (packet and control plane config) but BMv2 also models all the other parts of the stack such as the control plane API etc. So are a little bit more limited here.